File: cts/tests/tests/renderscript/src/android/renderscript/cts/IntrinsicBlur.java
/*
 * Copyright (C) 2013 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.renderscript.cts;

import android.renderscript.*;
import android.util.Log;

public class IntrinsicBlur extends IntrinsicBase {
    private ScriptIntrinsicBlur mIntrinsic;
    private int MAX_RADIUS = 25;
    private ScriptC_intrinsic_blur mScript;
    private float mRadius = MAX_RADIUS;
    private float mSaturation = 1.0f;
    private Allocation mScratchPixelsAllocation1;
    private Allocation mScratchPixelsAllocation2;



    private void initTest(int w, int h, Element e, Script.LaunchOptions lo) {
        makeBuffers(w, h, e);

        Type.Builder tb = new Type.Builder(mRS, Element.F32_4(mRS));
        tb.setX(w);
        tb.setY(h);
        mScratchPixelsAllocation1 = Allocation.createTyped(mRS, tb.create());
        mScratchPixelsAllocation2 = Allocation.createTyped(mRS, tb.create());

        mIntrinsic = ScriptIntrinsicBlur.create(mRS, e);
        mIntrinsic.setRadius(MAX_RADIUS);
        mIntrinsic.setInput(mAllocSrc);

        mScript = new ScriptC_intrinsic_blur(mRS);
        mScript.set_width(w);
        mScript.set_height(h);
        mScript.invoke_setRadius(MAX_RADIUS);

        mScript.set_ScratchPixel1(mScratchPixelsAllocation1);
        mScript.set_ScratchPixel2(mScratchPixelsAllocation2);

        // Make reference
        copyInput();
        mScript.forEach_horz(mScratchPixelsAllocation2);
        mScript.forEach_vert(mScratchPixelsAllocation1);
        copyOutput(lo);
    }

    private void copyInput() {
        if (mAllocSrc.getType().getElement().isCompatible(Element.U8(mRS))) {
            mScript.forEach_convert1_uToF(mAllocSrc, mScratchPixelsAllocation1);
            return;
        }
        if (mAllocSrc.getType().getElement().isCompatible(Element.U8_4(mRS))) {
            mScript.forEach_convert4_uToF(mAllocSrc, mScratchPixelsAllocation1);
            return;
        }
        throw new IllegalArgumentException("bad type");
    }

    private void copyOutput(Script.LaunchOptions lo) {
        if (mAllocSrc.getType().getElement().isCompatible(Element.U8(mRS))) {
            mScript.forEach_convert1_fToU(mScratchPixelsAllocation1, mAllocRef, lo);
            return;
        }
        if (mAllocSrc.getType().getElement().isCompatible(Element.U8_4(mRS))) {
            mScript.forEach_convert4_fToU(mScratchPixelsAllocation1, mAllocRef, lo);
            return;
        }
        throw new IllegalArgumentException("bad type");
    }

    public void testU8_1() {
        final int w = 97;
        final int h = 97;
        Element e = Element.U8(mRS);
        initTest(w, h, e, null);

        mIntrinsic.forEach(mAllocDst);

        mVerify.set_gAllowedIntError(1);
        mVerify.invoke_verify(mAllocRef, mAllocDst, mAllocSrc);
        mRS.finish();
        checkError();
    }

    public void testU8_4() {
        final int w = 97;
        final int h = 97;
        Element e = Element.U8_4(mRS);
        initTest(w, h, e, null);

        mIntrinsic.forEach(mAllocDst);

        mVerify.set_gAllowedIntError(1);
        mVerify.invoke_verify(mAllocRef, mAllocDst, mAllocSrc);
        mRS.finish();
        checkError();
    }


    public void testU8_1C() {
        final int w = 97;
        final int h = 97;
        Element e = Element.U8(mRS);
        Script.LaunchOptions lo = makeClipper(11, 11, w - 11, h - 11);

        initTest(w, h, e, lo);
        mIntrinsic.forEach(mAllocDst, lo);

        mVerify.set_gAllowedIntError(1);
        mVerify.invoke_verify(mAllocRef, mAllocDst, mAllocSrc);
        mRS.finish();
        checkError();
    }

    public void testU8_4C() {
        final int w = 97;
        final int h = 97;
        Element e = Element.U8_4(mRS);
        Script.LaunchOptions lo = makeClipper(11, 11, w - 11, h - 11);

        initTest(w, h, e, lo);
        mIntrinsic.forEach(mAllocDst, lo);

        mVerify.set_gAllowedIntError(1);
        mVerify.invoke_verify(mAllocRef, mAllocDst, mAllocSrc);
        mRS.finish();
        checkError();
    }

    public void test_ID() {
        ScriptIntrinsicBlur s = ScriptIntrinsicBlur.create(mRS, Element.U8_4(mRS));
        Script.KernelID kid = s.getKernelID();
        if (kid == null) {
            throw new IllegalStateException("kid must be valid");
        }

        Script.FieldID fid = s.getFieldID_Input();
        if (fid == null) {
            throw new IllegalStateException("fid must be valid");
        }
    }

}

File: cts/tests/tests/renderscript/src/android/renderscript/cts/intrinsic_blur.rs
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "shared.rsh"

int height;
int width;
static int radius;
rs_allocation ScratchPixel1;
rs_allocation ScratchPixel2;

const int MAX_RADIUS = 25;

// Store our coefficients here
static float gaussian[MAX_RADIUS * 2 + 1];

float4 __attribute__((kernel)) convert1_uToF(uchar v) {
    float4 r = rsUnpackColor8888(v);
    return r.r;
}

float4 __attribute__((kernel)) convert4_uToF(uchar4 v) {
    return rsUnpackColor8888(v);
}

uchar __attribute__((kernel)) convert1_fToU(float4 v) {
    uchar4 r = rsPackColorTo8888(v);
    return r.r;
}

uchar4 __attribute__((kernel)) convert4_fToU(float4 v) {
    return rsPackColorTo8888(v);
}

void setRadius(int rad) {
    // This function is a duplicate of:
    // RsdCpuScriptIntrinsicBlur::ComputeGaussianWeights()
    // Which is the reference C implementation
    radius = rad;
    const float e = M_E;
    const float pi = M_PI;
    float sigma = 0.4f * (float)radius + 0.6f;
    float coeff1 = 1.0f / (sqrt( 2.0f * pi ) * sigma);
    float coeff2 = - 1.0f / (2.0f * sigma * sigma);
    float normalizeFactor = 0.0f;
    float floatR = 0.0f;
    for (int r = -radius; r <= radius; r ++) {
        floatR = (float)r;
        gaussian[r + radius] = coeff1 * pow(e, floatR * floatR * coeff2);
        normalizeFactor += gaussian[r + radius];
    }

    normalizeFactor = 1.0f / normalizeFactor;
    for (int r = -radius; r <= radius; r ++) {
        floatR = (float)r;
        gaussian[r + radius] *= normalizeFactor;
    }
}

float4 __attribute__((kernel)) vert(uint32_t x, uint32_t y) {
    float4 blurredPixel = 0;
    int gi = 0;
    for (int r = -radius; r <= radius; r ++) {
        int validH = clamp((int)y + r, (int)0, (int)(height - 1));
        float4 i = rsGetElementAt_float4(ScratchPixel2, x, validH);
        blurredPixel += i * gaussian[gi++];
    }
    return blurredPixel;
}

float4 __attribute__((kernel)) horz(uint32_t x, uint32_t y) {
    float4 blurredPixel = 0;
    int gi = 0;
    for (int r = -radius; r <= radius; r ++) {
        // Stepping left and right away from the pixel
        int validX = clamp((int)x + r, (int)0, (int)(width - 1));
        float4 i = rsGetElementAt_float4(ScratchPixel1, validX, y);
        blurredPixel += i * gaussian[gi++];
    }
    return blurredPixel;
}
